numba_typing/overload_list.py

```diff
@@ -0,0 +1,183 @@
+        if isinstance(p_type, type):
+            return res
+        for p_val, n_val in zip(p_type.__args__, n_type.key):
+            res = res and self.match(p_val, n_val)
+    if isinstance(n_type, types.UniTuple):
+        res = True
+        if isinstance(p_type, type):
+            return res
+        for p_val in p_type.__args__:
+            res = res and self.match(p_val, n_type.key[0])
+    return res
+
+
+def check_dict_type(self, p_type, n_type):
+    res = False
+    if isinstance(n_type, types.DictType):
+        res = True
+        if isinstance(p_type, type):
+            return res
+        for p_val, n_val in zip(p_type.__args__, n_type.keyvalue_type):
+            res = res and self.match(p_val, n_val)
+    return res
+
+
+class TypeChecker:
+
+    _types_dict = {int: check_int_type, float: check_float_type, bool: check_bool_type,
+                   str: check_str_type, list: check_list_type,
+                   tuple: check_tuple_type, dict: check_dict_type}
+
+    def __init__(self):
+        self._typevars_dict = {}
+
+    def clear_typevars_dict(self):
+        self._typevars_dict.clear()
+
+    def add_type_check(self, type_check, func):
+        self._types_dict[type_check] = func
+
+    def _is_generic(self, p_obj):
+        if isinstance(p_obj, typing._GenericAlias):
+            return True
+
+        if isinstance(p_obj, typing._SpecialForm):
+            return p_obj not in {typing.Any}
+
+        return False
+
+    def _get_origin(self, p_obj):
+        return p_obj.__origin__
+
+    def match(self, p_type, n_type):
+        try:
+            if p_type == typing.Any:
+                return True
+            elif self._is_generic(p_type):
+                origin_type = self._get_origin(p_type)
+                if origin_type == typing.Generic:
+                    return self.match_generic(p_type, n_type)
+                else:
+                    return self._types_dict[origin_type](self, p_type, n_type)
+            elif isinstance(p_type, typing.TypeVar):
+                return self.match_typevar(p_type, n_type)
+            else:
+                if p_type in (list, tuple):
+                    return self._types_dict[p_type](self, p_type, n_type)
+                return self._types_dict[p_type](n_type)
+        except KeyError:
+            print((f'A check for the {p_type} was not found. {n_type}'))
+
+    def match_typevar(self, p_type, n_type):
+        if not self._typevars_dict.get(p_type) and n_type not in self._typevars_dict.values():
+            self._typevars_dict[p_type] = n_type
+            return True
+        return self._typevars_dict.get(p_type) == n_type
+
+    def match_generic(self, p_type, n_type):
+        res = True
+        for arg in p_type.__args__:
+            res = res and self.match(arg, n_type)
+        return res
+
+
+def choose_func_by_sig(sig_list, values_dict, defaults_dict={}):
+    checker = TypeChecker()
+    for sig in sig_list:  # sig = (Signature,func)
```

Comment:
Instead of accessing `sig[0]` and `sig[1]` each time you can write `for sig, func in sig_list:`